### DISCUSSION TITLE
Custom type of network

### DISCUSSION THREAD

**Discussion by EvgenyUgolkov** [PMEAL/OpenPNM#2682] [Q&A] [answered]
Dear Team, good day!

Could you consult me, is it possible to create a custom type of network? What i mean by that. I have 3D Micro-CT image, and get a network from it with the PoreSpy as follows:

**_snow_output = ps.networks.snow2(my_image, my_voxel_size)_**

Then, i take this network and put it into the Openpnm as follows:

**_pn = op.io.network_from_porespy(snow_output.network)_**

This way, if i understand it correct, now i have a topology, but the shapes of pores and throats are not defined yet. On this stage, i can use 

**_modules - models - collections - geometry_** - and choose any "ready" geometry we have (for example, circles_and_rectangles, etc)

For example, i can do

**_pn.add_model_collection(op.models.collections.geometry.spheres_and_cylinders)_**

Here is my question. If i want to assign my own custom shape of pores and throats, then what syntax should i use and do i have any choice? For example, pyramid pores and cuboid throats? Could you please provide an example?

With respect, Evgeny

**Comment by jgostick** [comment, marked as answer]
Hi @EvgenyUgolkov
This is a good question.  When you do a snow2 extraction it includes numerous geometrical properties, like 'pore.inscribed_diameter', 'pore.region_volume', etc.  

The first thing to note is that we no longer assume what the user wants to be as their official 'pore.diameter' in OpenPNM.  You must assign your own, like ``pn['pore.diameter'] = pn['pore.inscribed_diameter']``.  

THEN, and this gets to your point, you need to make an assumption about what shapes to use. For instance, if you want to assume spherical pores, then you would do either of the following:
- ``pn['pore.volume'] = 3.14159/4*pn['pore.diameter']**3``
- ``pn.add_model(propname='pore.diameter', model=op.models.geometry.pore_volume.spheres)``

In the 2nd case you are using a prewritten function which does essentially what is done in the previous line. 

SO, if you want to create your own volume, you can either do:
- ``pn['pore.volume'] = <some complex function>``
- ``pn.add_model(propname='pore.diameter', model=<some custom pore-scale model you have written>)``

The advantage of the pore-scale model is that it allows you to update your pore volumes just by doing ``pn.regenerate_models()`` if the pore diameter changes for some reason. It also allows you to hide your volume calculation out of sight instead of in your main script.

[Here is an example](https://openpnm.org/examples/reference/architecture/pore_scale_models.html) on the details of creating your own pore-scale model.

Lastly, you'll probably want to create some transport conductance models which account for the shape you wish to create.  This can get considerably more complicated.  We have started using the concept of 'size factors', which are [explained here](https://openpnm.org/examples/reference/simulations/size_factors_and_transport_conductance.html). As you cand see, this will be a bit of work if you want to use something oddly shaped.  

Anyway, I hope this is enough to get you started.

**Reply by EvgenyUgolkov** [reply]
Thank you for the perfect explanation!

**Reply by fsamaraie** [reply]
Hello Dr @jgostick 
I would apperciate if you can clarify an issue for me. Once I use pn = op.io.network_from_porespy(snow_output.network) I get a model which already have an equivalent_diameter and an inscribed_diameter, as you explained above. However, how was the equivalent diameter counted here? by assuming a spherical pore or a cubic pore?
Thanks alot

**Reply by jgostick** [reply]
You can always look at the code base to answer things like this...but its based on spheres.

**Reply by fsamaraie** [reply]
Many thanks.
